scripts/helper_functions.sh

```diff
@@ -93,96 +96,192 @@ export YellowBoldText='\033[1;33m'      # Yellow
+    DiscordMessage "${DISCORD_POST_UPDATE_BOOT_MESSAGE}" "success"
 }
 
 # Returns 0 if game is installed
 # Returns 1 if game is not installed
 IsInstalled() {
-  if  [ -e /palworld/PalServer.sh ] && [ -e /palworld/steamapps/appmanifest_2394010.acf ]; then
-    return 0
-  fi
-  return 1
-}
\ No newline at end of file
+    if  [ -e /palworld/PalServer.sh ] && [ -e /palworld/steamapps/appmanifest_2394010.acf ]; then
+        return 0
+    fi
+    return 1
+}
+
+# Saves the server
+# Returns 0 if it saves
+# Returns 1 if it is not able to save
+save_server() {
+    local return_val=0
+    if ! RCON save; then
+        return_val=1
+    fi
+    return "$return_val"
+}
+
+# Saves then shutdowns the server
+# Returns 0 if it is shutdown
+# Returns 1 if it is not able to be shutdown
+shutdown_server() {
+    local return_val=0
+    # Do not shutdown if not able to save
+    if save_server; then
+        if ! RCON "DoExit"; then
+            return_val=1
+        fi
+    else
+        return_val=1
+    fi
+    return "$return_val"
+}
+
+# Given a message this will broadcast in game
+# Since RCON does not support spaces this will replace all spaces with underscores
+# Returns 0 on success
+# Returns 1 if not able to broadcast
+broadcast_command() {
+    local return_val=0
+    # Replaces spaces with underscore
+    local message="${1// /_}"
+    if ! RCON "broadcast ${message}"; then
+        return_val=1
+    fi
+    return "$return_val"
+}
+
+# Given an amount of time in minutes and a message prefix
+# Returns 0 on success
+# If mtime is not an integer and there are players in game then return 1
+countdown_message() {
+    local mtime="$1"
+    local message_prefix="$2"
+    local return_val=0
+
+    if [[ "${mtime}" =~ ^[0-9]+$ ]]; then
+        for ((i = "${mtime}" ; i > 0 ; i--)); do
+            if [ "$(get_player_count)" -eq 0 ]; then
+                break
+            fi
+            if [ "$i" -eq 1 ]; then
+                broadcast_command "${message_prefix}_in_${i}_minute"
+                sleep 30s
+                broadcast_command "${message_prefix}_in_30_seconds"
+                sleep 20s
+                broadcast_command "${message_prefix}_in_10_seconds"
+                sleep 10s
+            else
+                case "$i" in
+                    "$mtime" )
+                        ;&
+                    15 )
+                        ;&
+                    10 )
+                        ;&
+                    5 )
+                        ;&
+                    2 )
+                        broadcast_command "${message_prefix}_in_${i}_minutes"
+                        ;&
+                    * ) 
+                        sleep 1m
+                        ;;
+                esac
+            fi
+        done
+        return 0
+    else
+        # mtime is not an integer so check if there are no players
+        if [ "$(get_player_count)" -gt 0 ]; then
+            return_val=1
+        fi
+    fi
```

Comment:
I would suggest to negate `if [[ "${mtime}" =~ ^[0-9]+$ ]]; then` to reduce nested code.

Reply:
I have that because the only time this is called is during a shutdown for update or reboot.
I believe both of those should happen regardless if there's a timer or not as the timer is just to let players know. If there is no timer set and there are no players I believe the action should proceed.